teachertool/src/types/criteriaParameters.ts

```diff
@@ -0,0 +1,139 @@
+    default: string | undefined;
+    paths: string[]; // The json path(s) to update with the parameter value in the catalog criteria.
+
+    constructor(name: string, type: CriteriaParameterType, paths: string[], defaultValue?: string) {
+        this.name = name;
+        this.type = type;
+        this.default = defaultValue;
+        this.paths = paths;
+    }
+
+    validate(value: string): CriteriaParameterValidationResult {
+        return { valid: true };
+    }
+}
+
+export class BaseStringParameter extends CriteriaParameter {
+    maxCharacters: number | undefined;
+
+    constructor(
+        name: string,
+        type: CriteriaParameterType,
+        paths: string[],
+        defaultValue?: string,
+        maxCharacters?: number
+    ) {
+        super(name, type, paths, defaultValue);
+        this.maxCharacters = maxCharacters;
+    }
+
+    override validate(value: string): CriteriaParameterValidationResult {
+        if (!value) return { valid: true }; // Unset is okay for initial value
+
+        if (this.maxCharacters && value.length > this.maxCharacters) {
+            return { valid: false, message: Strings.ExceedsMaxLength };
+        }
+        return { valid: true };
+    }
+}
+
+export class StringParameter extends BaseStringParameter {
+    constructor(template: CriteriaParameter, maxCharacters?: number) {
+        super(template.name, "string", template.paths, template.default, maxCharacters);
+    }
+}
+
+export class LongStringParameter extends BaseStringParameter {
+    constructor(template: CriteriaParameter, maxCharacters?: number) {
+        super(template.name, "longString", template.paths, template.default, maxCharacters);
+    }
+}
+
+export class NumberParameter extends CriteriaParameter {
+    min: number | undefined;
+    max: number | undefined;
+
+    constructor(template: CriteriaParameter, min?: number, max?: number) {
+        super(template.name, "number", template.paths, template.default);
+        this.min = min;
+        this.max = max;
+    }
+
+    override validate(value: any): CriteriaParameterValidationResult {
+        // Ensure the value is numeric and within the specified range.
+        const num = Number(value);
+        if (isNaN(num)) {
+            return {
+                valid: false,
+                message: Strings.MustBeANumber,
+            };
+        }
+        if (this.min && num < this.min) {
+            return {
+                valid: false,
+                message: Strings.BelowMin,
+            };
+        }
+        if (this.max && num > this.max) {
+            return {
+                valid: false,
+                message: Strings.ExceedsMax,
+            };
+        }
+        return { valid: true };
+    }
+}
+
+export class BlockParameter extends CriteriaParameter {
+    constructor(template: CriteriaParameter) {
+        super(template.name, "block", template.paths, template.default);
+    }
+}
+
+export class SystemParameter extends CriteriaParameter {
```

Comment:
Just wondering, what is a SystemParameter? Maybe a doc comment would be helpful here.

Reply:
Added a comment to explain, but basically, it's a variable value that is set by the system rather than the user. The main scenario for this currently is when we pass the share id as a parameter to the AI criteria. There's no need for the user to enter that into the criteria instance since we have that data already in the state, so we just auto-populate it behind the scenes at eval time.